File: shared/src/types/product.ts
import type { CompetitionLevel, MarketplaceChannel, ProductStatus } from './enums';
import type { MarketingCampaignPlan, TrendScore } from './marketing';

/**
 * Atributos extraídos pela visão computacional (Agente 1).
 * Todos opcionais: uma foto pode não revelar tudo.
 */
export interface ProductVisionAttributes {
  name?: string;
  brand?: string;
  model?: string;
  category?: string;
  subcategory?: string;
  color?: string;
  material?: string;
  size?: string;
  barcode?: string;
  ean?: string;
  sku?: string;
  packageText?: string;
  quantity?: number;
  supplier?: string;
  labelPrice?: number; // preço da etiqueta, se visível
  shortDescription?: string;
  features?: string[];
  /** Peso em kg com embalagem. Obrigatório na Shopee — define o frete. */
  weight?: number;
  /**
   * Medidas do pacote em cm. A Shopee exige as três juntas — sem elas o produto
   * não é salvo ("Este campo não pode ficar em branco").
   */
  length?: number;
  width?: number;
  height?: number;
  /**
   * De onde saiu o `weight`. `etiqueta` foi lido da foto; `estimado` a IA
   * deduziu de material/tamanho/quantidade — não é medição. Serve para achar
   * os produtos a reconferir se o frete cobrado vier errado.
   */
  weightSource?: 'etiqueta' | 'estimado';
  /**
   * ID numérico da árvore de categorias da Shopee (ex.: 120039), digitado
   * pelo operador. A IA só extrai um caminho de texto ("Beleza > Maquiagem");
   * tanto a planilha de importação em massa quanto o Add/Update Item da API
   * (aqui obrigatório) precisam do ID — nunca inventado.
   */
  shopeeCategoryId?: number;
}

/** Resultado de pesquisa de mercado (Agente 2). */
export interface MarketResearchResult {
  averagePrice: number;
  minPrice: number;
  maxPrice: number;
  approxListingCount: number;
  marketRange: { from: number; to: number };
  similarProducts: Array<{
    title: string;
    price: number;
    source: string;
    url?: string;
  }>;
  competition: CompetitionLevel;
  sources: string[]; // marketplaces consultados
  collectedAt: string; // ISO
}

/** Conteúdo comercial/SEO gerado (Agente 3). */
export interface GeneratedContent {
  title: string;
  description: string;
  longDescription: string;
  summary: string;
  bulletPoints: string[];
  seo: {
    metaDescription: string;
    slug: string;
    keywords: string[];
    tags: string[];
  };
  category: string;
  technicalSpecs: Record<string, string>;
  marketplaceDescription: string; // versão curta p/ marketplace
}

/** Preço calculado (Agente 5). */
export interface PricingResult {
  purchasePrice: number;
  suggestedPrice: number; // já com "preço psicológico"
  markupApplied: number; // percentual (ex.: 0.9 = 90%)
  profit: number;
  marginPercent: number;
  roi: number;
}

/** Uma imagem tratada e suas variantes (Agente 4). */
export interface ProductImageSet {
  original: string; // URL storage
  hd?: string;
  square?: string;
  webp?: string;
  thumbnail?: string;
  backgroundRemoved?: string;
  isManufacturerProvided?: boolean;
  /** 3 imagens prontas p/ Shopee (1:1, fundo limpo): produto recortado e recomposto. */
  shopee?: string[];
}

/**
 * Rascunho de post social (Facebook/Instagram) aguardando aprovação do dono
 * via Telegram. Um produto tem no máximo um `socialApproval` ativo por vez —
 * se rejeitado, o próximo ciclo do agendador pode tentar outro produto.
 */
export interface SocialApproval {
  status: 'pending' | 'approved' | 'rejected' | 'posted';
  caption: string;
  telegramChatId: string;
  telegramMessageId: number;
  scheduledAt: string; // ISO — quando foi mandado pra aprovação
  postedAt?: string; // ISO — quando publicou de fato
}

/**
 * Agregado de produto — o documento central do catálogo.
 * `id` é o identificador do domínio; no Mongo mapeamos para `_id`.
 */
export interface Product {
  id: string;
  ownerId: string; // usuário dono do cadastro
  internalSku: string;
  status: ProductStatus;
  aiConfidence: number; // 0..1
  vision: ProductVisionAttributes;
  market?: MarketResearchResult;
  content?: GeneratedContent;
  pricing?: PricingResult;
  images: ProductImageSet;
  /** Quando a visão detecta vários produtos numa foto. */
  multipleProductsDetected?: boolean;
  publishedChannels: MarketplaceChannel[];
  /** ID do anúncio em cada canal externo (ex.: `{ shopee: "2201..." }`) — necessário para update/unlist. */
  externalIds?: Partial<Record<MarketplaceChannel, string>>;
  socialApproval?: SocialApproval;
  /** Análise do Marketing IA (Trend Hunter + Marketing Planner) — ver `@tecnoplus/shared/types/marketing`. */
  marketing?: {
    trend?: TrendScore;
    plan?: MarketingCampaignPlan;
  };
  createdAt: string;
  updatedAt: string;
}

File: backend/app/shared/schemas/product.py
from enum import StrEnum

from pydantic import BaseModel, ConfigDict, Field

from app.shared.schemas.enums import CompetitionLevel, MarketplaceChannel, ProductStatus
from app.shared.schemas.marketing import MarketingCampaignPlan, TrendScore


class WeightSource(StrEnum):
    ETIQUETA = "etiqueta"
    ESTIMADO = "estimado"


class ProductVisionAttributes(BaseModel):
    name: str | None = None
    brand: str | None = None
    model: str | None = None
    category: str | None = None
    subcategory: str | None = None
    color: str | None = None
    material: str | None = None
    size: str | None = None
    barcode: str | None = None
    ean: str | None = None
    sku: str | None = None
    packageText: str | None = None
    quantity: int | None = None
    supplier: str | None = None
    labelPrice: float | None = None
    shortDescription: str | None = None
    features: list[str] | None = None
    weight: float | None = None
    length: float | None = None
    width: float | None = None
    height: float | None = None
    weightSource: WeightSource | None = None
    shopeeCategoryId: int | None = None
    mercadoLivreCategoryId: str | None = None
    mercadoLivreListingTypeId: str | None = None


class SimilarProduct(BaseModel):
    title: str
    price: float
    source: str
    url: str | None = None


class MarketRange(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    from_: float = Field(alias="from")
    to: float


class MarketResearchResult(BaseModel):
    averagePrice: float
    minPrice: float
    maxPrice: float
    approxListingCount: int
    marketRange: MarketRange
    similarProducts: list[SimilarProduct]
    competition: CompetitionLevel
    sources: list[str]
    collectedAt: str


class SeoContent(BaseModel):
    metaDescription: str
    slug: str
    keywords: list[str]
    tags: list[str]


class GeneratedContent(BaseModel):
    title: str
    description: str
    longDescription: str
    summary: str
    bulletPoints: list[str]
    seo: SeoContent
    category: str
    technicalSpecs: dict[str, str]
    marketplaceDescription: str


class PricingResult(BaseModel):
    purchasePrice: float
    suggestedPrice: float
    markupApplied: float
    profit: float
    marginPercent: float
    roi: float


class ProductImageSet(BaseModel):
    original: str
    hd: str | None = None
    square: str | None = None
    webp: str | None = None
    thumbnail: str | None = None
    backgroundRemoved: str | None = None
    isManufacturerProvided: bool | None = None
    shopee: list[str] | None = None


class SocialApprovalStatus(StrEnum):
    PENDING = "pending"
    APPROVED = "approved"
    REJECTED = "rejected"
    POSTED = "posted"


class SocialApproval(BaseModel):
    status: SocialApprovalStatus
    caption: str
    telegramChatId: str
    telegramMessageId: int
    scheduledAt: str
    postedAt: str | None = None


class ProductMarketing(BaseModel):
    trend: TrendScore | None = None
    plan: MarketingCampaignPlan | None = None


class Product(BaseModel):
    id: str
    ownerId: str
    internalSku: str
    status: ProductStatus
    aiConfidence: float
    vision: ProductVisionAttributes
    market: MarketResearchResult | None = None
    content: GeneratedContent | None = None
    pricing: PricingResult | None = None
    images: ProductImageSet
    multipleProductsDetected: bool | None = None
    publishedChannels: list[MarketplaceChannel]
    externalIds: dict[MarketplaceChannel, str] | None = None
    socialApproval: SocialApproval | None = None
    marketing: ProductMarketing | None = None
    createdAt: str
    updatedAt: str
